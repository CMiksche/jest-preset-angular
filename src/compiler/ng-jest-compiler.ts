import { CompilerHost, CompilerOptions, createCompilerHost, Diagnostic } from '@angular/compiler-cli';
import { hasErrors } from '@ngtools/webpack/src/diagnostics';
import type { Logger } from 'bs-logger';
import { normalize } from 'path';
import { updateOutput } from 'ts-jest/dist/compiler/instance';
import { LINE_FEED } from 'ts-jest/dist/constants';
import type { TTypeScript } from 'ts-jest/dist/types';
import type * as ts from 'typescript';

import type { NgJestConfig } from '../config/ng-jest-config';
import { factory as downlevelCtor } from '../transformers/downlevel-ctor';

export class NgJestCompiler {
  private _compilerOptions!: CompilerOptions;
  private _program: ts.Program | undefined;
  private _compilerHost!: CompilerHost;
  private _rootNames: string[] = [];
  private _allDiagnostics: Array<ts.Diagnostic | Diagnostic> = [];
  private readonly _sourceFileCache: Map<string, ts.SourceFile> = new Map<string, ts.SourceFile>();
  private readonly _memoryHost: Map<string, string> = new Map<string, string>();
  private readonly _logger: Logger;
  private readonly _ts: TTypeScript;

  constructor(private readonly ngJestConfig: NgJestConfig) {
    this._logger = this.ngJestConfig.logger;
    this._ts = this.ngJestConfig.compilerModule;
    this._setupOptions(this.ngJestConfig);

    this._logger.debug('created NgJestCompiler');
  }

  getCompiledOutput(fileName: string, fileContent: string): string {
    if (this._program) {
      if (!this._rootNames.includes(fileName)) {
        this._logger.debug({ fileName }, 'getCompiledOutput: update memory host, rootFiles and Program');

        this._rootNames = [...this._rootNames, fileName];
        this._memoryHost.set(fileName, fileContent);
        this._createOrUpdateProgram();
      }
      this._logger.debug({ fileName }, 'getCompiledOutput: compiling using Program');

      const compiledOutput: string[] = [];
      const sourceFile = this._program.getSourceFile(fileName);
      const emitResult = this._program.emit(
        sourceFile,
        (_fileName: string, data: string, _writeByteOrderMark: boolean) => {
          compiledOutput.push(data);
        },
        undefined,
        undefined,
        {
          before: [
            // hoisting from `ts-jest` or other before transformers
            ...(this.ngJestConfig.customTransformers.before as ts.TransformerFactory<ts.SourceFile>[]),
            /**
             * Downlevel constructor parameters for DI support. This is required to support forwardRef in ES2015 due to
             * TDZ issues. This wrapper is needed here due to the program not being available until after
             * the transformers are created. Also because program can be updated so we can't push this transformer in
             * _createCompilerHost
             */
            downlevelCtor(this._program),
          ],
        },
      );
      if (this.ngJestConfig.shouldReportDiagnostics(fileName)) {
        this._logger.debug({ fileName }, 'getCompiledOutput: getting diagnostics');

        this._allDiagnostics.push(
          ...emitResult.diagnostics,
          ...this._program.getSyntacticDiagnostics(sourceFile),
          ...this._program.getSemanticDiagnostics(sourceFile),
        );
      }
      if (!hasErrors(this._allDiagnostics)) {
        this._logger.debug({ fileName }, 'getCompiledOutput: update compiled output including inline source map');

        return updateOutput(compiledOutput[1], fileName, compiledOutput[0]);
      } else {
        this.ngJestConfig.raiseDiagnostics(this._allDiagnostics as ts.Diagnostic[], fileName);

        return '';
      }
    } else {
      this._logger.debug({ fileName }, 'getCompiledOutput: compiling as isolated module');

      const result: ts.TranspileOutput = this._ts.transpileModule(fileContent, {
        fileName,
        transformers: this.ngJestConfig.customTransformers,
        compilerOptions: this._compilerOptions,
        reportDiagnostics: this.ngJestConfig.shouldReportDiagnostics(fileName),
      });
      if (result.diagnostics && this.ngJestConfig.shouldReportDiagnostics(fileName)) {
        this.ngJestConfig.raiseDiagnostics(result.diagnostics, fileName, this._logger);
      }

      // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
      return updateOutput(result.outputText, fileName, result.sourceMapText!);
    }
  }

  private _setupOptions({ parsedTsConfig }: NgJestConfig) {
    this._logger.debug({ parsedTsConfig }, '_setupOptions: initializing compiler config');

    this._compilerOptions = { ...parsedTsConfig.options };
    this._rootNames = parsedTsConfig.rootNames.filter((rootName) => !this.ngJestConfig.isTestFile(rootName));
    if (this._compilerOptions.strictMetadataEmit) {
      this._logger.warn(
        `Using Angular compiler option 'strictMetadataEmit' for applications might cause undefined behavior.`,
      );
    }
    if (!this.ngJestConfig.isolatedModules) {
      this._logger.debug(
        { compilerOptions: this._compilerOptions },
        '_setupOptions: creating Compiler Host using @angular/compiler-cli createCompilerHost',
      );

      this._createCompilerHost();
      this._createOrUpdateProgram();
    }
  }

  private _createCompilerHost(): void {
    this._logger.debug('_createCompilerHost: creating compiler host using @angular/compiler-cli');

    this._compilerHost = createCompilerHost({
      options: this._compilerOptions,
    });
    this._compilerHost.readFile = (fileName: string): string | undefined => {
      const normalizedFileName = normalize(fileName);
      let fileContent = this._memoryHost.get(normalizedFileName);
      if (!fileContent) {
        fileContent = this._ts.sys.readFile(normalizedFileName, 'utf-8') ?? '';
        this._memoryHost.set(normalizedFileName, fileContent);
      }

      return fileContent;
    };
    this._compilerHost.getSourceFile = (fileName: string, languageVersion: ts.ScriptTarget) => {
      const normalizedFileName = normalize(fileName);
      try {
        const cached = this._sourceFileCache.get(normalizedFileName);
        if (cached) {
          return cached;
        }

        this._logger.debug(
          { normalizedFileName },
          'getSourceFile: cache miss, reading file and create source file to update cache',
        );

        const content = this._compilerHost.readFile(normalizedFileName);
        if (content !== undefined) {
          const sf = this._ts.createSourceFile(normalizedFileName, content, languageVersion, true);
          this._sourceFileCache.set(fileName, sf);

          return sf;
        }
      } catch (e) {
        this._logger.error(e);
      }

      return undefined;
    };
    this._compilerHost.getCurrentDirectory = () => this.ngJestConfig.cwd;
    this._compilerHost.getNewLine = () => LINE_FEED;
    this._compilerHost.getDefaultLibFileName = (options: ts.CompilerOptions) => this._ts.getDefaultLibFilePath(options);
    this._compilerHost.trace = (message: string) => this._logger.trace(message);
  }

  private _createOrUpdateProgram(): void {
    const oldTsProgram = this._program;

    this._logger.debug(`_createOrUpdateProgram: ${oldTsProgram ? 'update' : 'create'} TypeScript Program`);

    this._program = this._ts.createProgram(this._rootNames, this._compilerOptions, this._compilerHost, oldTsProgram);
  }
}
