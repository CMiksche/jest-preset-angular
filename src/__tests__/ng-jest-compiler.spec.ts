import { readFileSync } from 'fs';
import { join } from 'path';
import { SOURCE_MAPPING_PREFIX } from 'ts-jest/dist/compiler/instance';

import { NgJestCompiler } from '../compiler/ng-jest-compiler';
import { NgJestConfig } from '../config/ng-jest-config';

// eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
const jestCfgStub = {
  globals: {
    'ts-jest': {
      diagnostics: {
        pretty: false,
      },
    },
  },
} as any; // eslint-disable-line @typescript-eslint/no-explicit-any
const ngJestConfig = new NgJestConfig(jestCfgStub);

function executeHoistingTest() {
  // Verify if we use `ts-jest` hoisting transformer
  test('should hoist correctly', () => {
    const fileName = join(__dirname, '__mocks__', 'foo.spec.ts');
    ngJestConfig.parsedTsConfig = {
      ...ngJestConfig.parsedTsConfig,
      rootNames: [fileName],
    };
    const compiler = new NgJestCompiler(ngJestConfig);

    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
    const emitResult = compiler.getCompiledOutput(fileName, readFileSync(fileName, 'utf-8'))!;

    // Source map is different based on file location which can fail on CI, so we only compare snapshot for js
    expect(emitResult.substring(0, emitResult.indexOf(SOURCE_MAPPING_PREFIX))).toMatchSnapshot();
  });
}

describe('NgJestCompiler', () => {
  describe('with isolatedModules true', () => {
    // Isolated modules true doesn't have downlevel ctor so this snapshot test should produce different input than with Program
    test('should return result when using isolatedModules true', () => {
      const fileName = join(__dirname, '__mocks__', 'foo.service.ts');
      const compiler = new NgJestCompiler(
        new NgJestConfig({
          globals: {
            // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
            'ts-jest': {
              ...jestCfgStub.globals['ts-jest'],
              isolatedModules: true,
            },
          },
        } as any), // eslint-disable-line @typescript-eslint/no-explicit-any
      );

      // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
      const emitResult = compiler.getCompiledOutput(fileName, readFileSync(fileName, 'utf-8'))!;

      // Source map is different based on file location which can fail on CI, so we only compare snapshot for js
      expect(emitResult.substring(0, emitResult.indexOf(SOURCE_MAPPING_PREFIX))).toMatchSnapshot();
    });

    executeHoistingTest();
  });

  describe('with isolatedModule false', () => {
    test('should return compiled result for new file which is not known by Program', () => {
      const fileName = join(__dirname, '__mocks__', 'app.component.ts');
      ngJestConfig.parsedTsConfig = {
        ...ngJestConfig.parsedTsConfig,
        rootNames: [],
      };
      const compiler = new NgJestCompiler(ngJestConfig);

      // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
      const emitResult = compiler.getCompiledOutput(fileName, readFileSync(fileName, 'utf-8'))!;

      // Source map is different based on file location which can fail on CI, so we only compare snapshot for js
      expect(emitResult.substring(0, emitResult.indexOf(SOURCE_MAPPING_PREFIX))).toMatchSnapshot();
    });

    test('should return compiled result for existing file which is known by Program', () => {
      const fileName = join(__dirname, '__mocks__', 'app.component.ts');
      ngJestConfig.parsedTsConfig = {
        ...ngJestConfig.parsedTsConfig,
        rootNames: [fileName],
      };
      const compiler = new NgJestCompiler(ngJestConfig);

      // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
      const emitResult = compiler.getCompiledOutput(fileName, readFileSync(fileName, 'utf-8'))!;

      // Source map is different based on file location which can fail on CI, so we only compare snapshot for js
      expect(emitResult.substring(0, emitResult.indexOf(SOURCE_MAPPING_PREFIX))).toMatchSnapshot();
    });

    test('should throw diagnostics error for new file which is not known by Program', () => {
      const fileName = join(__dirname, '__mocks__', 'foo.component.ts');
      ngJestConfig.parsedTsConfig = {
        ...ngJestConfig.parsedTsConfig,
        rootNames: [],
      };
      const compiler = new NgJestCompiler(ngJestConfig);

      expect(() =>
        compiler.getCompiledOutput('foo.ts', readFileSync(fileName, 'utf-8')),
      ).toThrowErrorMatchingSnapshot();
    });

    test('should throw diagnostics error for existing file which is known by Program', () => {
      const fileName = join(__dirname, '__mocks__', 'foo.component.ts');
      ngJestConfig.parsedTsConfig = {
        ...ngJestConfig.parsedTsConfig,
        rootNames: [fileName],
      };
      const compiler = new NgJestCompiler(ngJestConfig);

      expect(() =>
        compiler.getCompiledOutput(fileName, readFileSync(fileName, 'utf-8')),
      ).toThrowErrorMatchingSnapshot();
    });

    test('should not throw diagnostics error when shouldReportDiagnostics return false', () => {
      const fileName = join(__dirname, '__mocks__', 'foo.component.ts');
      ngJestConfig.parsedTsConfig = {
        ...ngJestConfig.parsedTsConfig,
        rootNames: [fileName],
      };
      ngJestConfig.shouldReportDiagnostics = jest.fn().mockReturnValueOnce(false);
      const compiler = new NgJestCompiler(ngJestConfig);

      expect(() => compiler.getCompiledOutput(fileName, readFileSync(fileName, 'utf-8'))).not.toThrowError();
    });

    executeHoistingTest();

    /**
     * This test is copied from https://github.com/angular/angular/blob/master/packages/compiler-cli/test/transformers/downlevel_decorators_transform_spec.ts
     * Only one test is enough to verify that our NgJestCompiler does use Angular downlevel ctor transformer.
     */
    test('should downlevel decorators for @Injectable decorated class', () => {
      const fileName = join(__dirname, '__mocks__', 'foo.service.ts');
      ngJestConfig.parsedTsConfig = {
        ...ngJestConfig.parsedTsConfig,
        rootNames: [fileName],
      };
      const compiler = new NgJestCompiler(ngJestConfig);
      // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
      const emitResult = compiler.getCompiledOutput(fileName, readFileSync(fileName, 'utf-8'))!;

      // Source map is different based on file location which can fail on CI, so we only compare snapshot for js
      expect(emitResult.substring(0, emitResult.indexOf(SOURCE_MAPPING_PREFIX))).toMatchSnapshot();
    });
  });
});
