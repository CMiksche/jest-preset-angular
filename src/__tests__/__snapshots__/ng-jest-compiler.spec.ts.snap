// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NgJestCompiler with isolatedModule false should downlevel decorators for @Injectable decorated class 1`] = `
"\\"use strict\\";
Object.defineProperty(exports, \\"__esModule\\", { value: true });
exports.MyService = exports.ClassInject = void 0;
var tslib_1 = require(\\"tslib\\");
var core_1 = require(\\"@angular/core\\");
var ClassInject = /** @class */ (function () {
    function ClassInject() {
    }
    return ClassInject;
}());
exports.ClassInject = ClassInject;
var MyService = /** @class */ (function () {
    // eslint-disable-next-line
    function MyService(_v) {
    }
    MyService.ctorParameters = function () { return [
        { type: ClassInject }
    ]; };
    MyService = tslib_1.__decorate([
        core_1.Injectable()
    ], MyService);
    return MyService;
}());
exports.MyService = MyService;
//# "
`;

exports[`NgJestCompiler with isolatedModule false should hoist correctly 1`] = `
"\\"use strict\\";
Object.defineProperty(exports, \\"__esModule\\", { value: true });
jest.mock('./foo');
var foo_1 = require(\\"./foo\\");
console.log(foo_1.getFoo());
//# "
`;

exports[`NgJestCompiler with isolatedModule false should return compiled result for existing file which is known by Program 1`] = `
"\\"use strict\\";
Object.defineProperty(exports, \\"__esModule\\", { value: true });
exports.AppComponent = void 0;
var tslib_1 = require(\\"tslib\\");
var core_1 = require(\\"@angular/core\\");
var AppComponent = /** @class */ (function () {
    function AppComponent() {
        this.title = 'test-app-v10';
    }
    AppComponent = tslib_1.__decorate([
        core_1.Component({
            selector: 'app-root',
            templateUrl: './app.component.html',
            styleUrls: ['./app.component.scss'],
        })
    ], AppComponent);
    return AppComponent;
}());
exports.AppComponent = AppComponent;
//# "
`;

exports[`NgJestCompiler with isolatedModule false should return compiled result for new file which is not known by Program 1`] = `
"\\"use strict\\";
Object.defineProperty(exports, \\"__esModule\\", { value: true });
exports.AppComponent = void 0;
var tslib_1 = require(\\"tslib\\");
var core_1 = require(\\"@angular/core\\");
var AppComponent = /** @class */ (function () {
    function AppComponent() {
        this.title = 'test-app-v10';
    }
    AppComponent = tslib_1.__decorate([
        core_1.Component({
            selector: 'app-root',
            templateUrl: './app.component.html',
            styleUrls: ['./app.component.scss'],
        })
    ], AppComponent);
    return AppComponent;
}());
exports.AppComponent = AppComponent;
//# "
`;

exports[`NgJestCompiler with isolatedModule false should throw diagnostics error for existing file which is known by Program 1`] = `"src/__tests__/__mocks__/foo.component.ts(8,3): error TS2322: Type '\\"test-app-v10\\"' is not assignable to type 'number'."`;

exports[`NgJestCompiler with isolatedModule false should throw diagnostics error for new file which is not known by Program 1`] = `"foo.ts(8,3): error TS2322: Type '\\"test-app-v10\\"' is not assignable to type 'number'."`;

exports[`NgJestCompiler with isolatedModules true should hoist correctly 1`] = `
"\\"use strict\\";
Object.defineProperty(exports, \\"__esModule\\", { value: true });
jest.mock('./foo');
var foo_1 = require(\\"./foo\\");
console.log(foo_1.getFoo());
//# "
`;

exports[`NgJestCompiler with isolatedModules true should return result 1`] = `
"\\"use strict\\";
Object.defineProperty(exports, \\"__esModule\\", { value: true });
exports.MyService = exports.ClassInject = void 0;
var tslib_1 = require(\\"tslib\\");
var core_1 = require(\\"@angular/core\\");
var ClassInject = /** @class */ (function () {
    function ClassInject() {
    }
    return ClassInject;
}());
exports.ClassInject = ClassInject;
var MyService = /** @class */ (function () {
    // eslint-disable-next-line
    function MyService(_v) {
    }
    MyService = tslib_1.__decorate([
        core_1.Injectable()
    ], MyService);
    return MyService;
}());
exports.MyService = MyService;
//# "
`;
